internal/internal_schedule_client.go

```diff
@@ -0,0 +1,789 @@
+	scheduleClient struct {
+		workflowClient *WorkflowClient
+	}
+
+	// scheduleHandleImpl is the implementation of ScheduleHandle.
+	scheduleHandleImpl struct {
+		ID     string
+		client *WorkflowClient
+	}
+
+	// scheduleListIteratorImpl is the implementation of ScheduleListIterator
+	scheduleListIteratorImpl struct {
+		// nextScheduleIndex - Local cached schedules events and corresponding consuming index
+		nextScheduleIndex int
+
+		// err - From getting the latest page of schedules
+		err error
+
+		// response - From getting the latest page of schedules
+		response *workflowservice.ListSchedulesResponse
+
+		// paginate - Function which use a next token to get next page of schedules events
+		paginate func(nexttoken []byte) (*workflowservice.ListSchedulesResponse, error)
+	}
+)
+
+func (w *workflowClientInterceptor) CreateSchedule(ctx context.Context, in *ScheduleClientCreateInput) (ScheduleHandle, error) {
+	// This is always set before interceptor is invoked
+	ID := in.Options.ID
+	if ID == "" {
+		return nil, fmt.Errorf("no schedule ID in options")
+	}
+
+	dataConverter := WithContext(ctx, w.client.dataConverter)
+	if dataConverter == nil {
+		dataConverter = converter.GetDefaultDataConverter()
+	}
+
+	if in.Options.Action == nil {
+		return nil, fmt.Errorf("no schedule action in options")
+	}
+	action, err := convertToPBScheduleAction(ctx, w.client, in.Options.Action)
+	if err != nil {
+		return nil, err
+	}
+
+	memo, err := getWorkflowMemo(in.Options.Memo, dataConverter)
+	if err != nil {
+		return nil, err
+	}
+
+	searchAttr, err := serializeSearchAttributes(in.Options.SearchAttributes)
+	if err != nil {
+		return nil, err
+	}
+
+	var triggerImmediately *schedulepb.TriggerImmediatelyRequest
+	if in.Options.TriggerImmediately {
+		triggerImmediately = &schedulepb.TriggerImmediatelyRequest{
+			OverlapPolicy: in.Options.Overlap,
+		}
+	}
+
+	backfillRequests := convertToPBBackfillList(in.Options.ScheduleBackfill)
+
+	// Only send an initial patch if we need to.
+	var initialPatch *schedulepb.SchedulePatch
+	if in.Options.TriggerImmediately || len(in.Options.ScheduleBackfill) > 0 {
+		initialPatch = &schedulepb.SchedulePatch{
+			TriggerImmediately: triggerImmediately,
+			BackfillRequest:    backfillRequests,
+		}
+	}
+
+	// run propagators to extract information about tracing and other stuff, store in headers field
+	startRequest := &workflowservice.CreateScheduleRequest{
+		Namespace:  w.client.namespace,
+		ScheduleId: ID,
+		RequestId:  uuid.New(),
+		Schedule: &schedulepb.Schedule{
+			Spec:   convertToPBScheduleSpec(&in.Options.Spec),
+			Action: action,
+			Policies: &schedulepb.SchedulePolicies{
+				OverlapPolicy:  in.Options.Overlap,
+				CatchupWindow:  &in.Options.CatchupWindow,
+				PauseOnFailure: in.Options.PauseOnFailure,
+			},
+			State: &schedulepb.ScheduleState{
+				Notes:            in.Options.Note,
+				Paused:           in.Options.Paused,
+				LimitedActions:   in.Options.RemainingActions != 0,
```

Comment:
In describe and update, the value of `RemainingActions` has to be interpreted in regard of `LimitedActions`. Assuming `LimitedActions == true`, then `RemainingActions == 0` would indicate that the schedule can't happens anymore (ie. it is implicitely paused), until remainingActions is reset to a non zero value.

Having only `RemainingActions` in create might make this semantic harder to understand for users. I'd suggest either adding `LimitedActions` to create options, or using a value other than 0 (ie `-1`) for "don't limit on action count", and using that same semantic in Describe and Update.